Image builder documentation (#9839)
kyma-project/test-infra: cmd/image-builder/README.md

# image-builder

Image Builder is a tool for building OCI-compliant images.

> It can build images using different backends, such as Kaniko, Buildkit, and Azure DevOps (ADO).

Reviewer: ```suggestion
It can build images using different backends, such as Kaniko, BuildKit, and Azure DevOps (ADO).
```

              - "--config=/config/kaniko-build-config.yaml"
              - "--context=."
              - "--dockerfile=cmd/image-builder/images/buildkit/Dockerfile"
              - "--build-in-ado=true"
            resources:
              requests:
                memory: 500Mi
                cpu: 500m
            volumeMounts:
              - name: config
                mountPath: /config
                readOnly: true
        volumes:
          - name: config
            configMap:
              name: kaniko-build-config
```

It builds the `buildkit-image-builder` image using the image-builder ADO backend.
It uses the Dockerfile from the `cmd/image-builder/images/buildkit/Dockerfile` path and the config from the `kaniko-build-config` ConfigMap.
Because it's a presubmit ProwJob it does not sign the image.
Signing images is supported only in postsubmit ProwJobs.

## Configuration

Image Builder is configured using a global configuration YAML file, set of environment variables, and command line flags.

### Configuration YAML File

`image-builder` requires a configuration YAML file. The file holds the global configuration for the tool and is maintained by the authors.

> Use `--config` flag to provide a path to the config YAML file.

Reviewer: ```suggestion
Use the `--config` flag to provide a path to the config YAML file.
```

For more information about available properties in the configuration file, refer to the [config.go](config.go) file.

> Example file:

Reviewer: ```suggestion
Here's an example file:
```

> # image-builder

Reviewer: Within the doc, you use image-builder and `image-builder`. I suggest using Image Builder as a proper name and `image-builder` for code-related contexts.

> Currently, image-builder contains a basic implementation of a notary signer. If you want to add a new signer, refer to

Reviewer: ```suggestion
Currently, Image Builder contains a basic implementation of a notary signer. If you want to add a new signer, refer to
```

Reviewer: There's one more "image-builder" in line 155. Please change to "Image Builder".

### Sign-Only Mode

Image Builder supports sign-only mode. To enable it, use the `--sign-only` flag.
It signs the images provided in the `--images-to-sign` flag.
It supports signing multiple images at once. The flag can be used multiple times.

## Named Tags

Image Builder supports passing the name along with the tag, using both the `-tag` option and the config for the tag template.
You can use `-tag name=value` to pass the name for the tag. 

If the name is not provided, it is evaluated from the value:
 - if the value is a string, it is used as a name directly. For example,`-tag latest` is equal to `-tag latest=latest`
 - if the value is go-template, it will be converted to a valid name. For example, `-tag v{{ .ShortSHA }}-{{ .Date }}` is equal to `-tag vShortSHA-Date=v{{ .ShortSHA }}-{{ .Date }}`

### Parse-Tags-Only Mode

You can use Image Builder to generate tags using pars-tags-only mode. To enable it, use the `--parse-tags-only` flag.
It parses the tags provided in the `--tag` flag and in `config.yaml`. The generated tags are written as JSON to
stdout.

## Build Backend

Image Builder supports three build backends:

- kaniko

> - Buildkit

Reviewer: ```suggestion
- BuildKit
```

Buildkit and Kaniko backends are deprecated and will be removed in the future.
The Preferred way to build images is to use ado backend, because it's the only SLC-29 compliant backend.

### Azure devops backend

Azure devops backend uses image-builder to call ADO API and trigger oci-image-builder pipeline. This backend is SLC-29
compliant. It supports signing images with production signify service. Images build with ADO can be pushed into kyma GCP
artifacts registers. To build images ADO backend uses `kaniko-project/executor` image. This backend doesn't
support `--env-file`, `--platform` and `--variant` flags. Building images for platforms other than amd64 is not
supported. To use this backend, you need to use image-builder in a prowjob. See [Quickstart guide](#quickstart-guide)
for
example prowjob definition.

When using ADO backend, an image-builder is used as client collecting values from flags and environment variables and
calling ADO API. Image-builder triggers oci-image-builder pipeline. This pipeline is responsible for processing
parameters provided in a call and building, pushing and signing image.

> Apart from calling ADO API to trigger image build, image-builder also supports preview mode. In preview mode,
> image-builder will not trigger ADO pipeline but will generate yaml file with pipeline definition. Using this mode allows
> to validate syntax of pipeline definition before triggering it. To use preview mode use `--ado-preview-run=true` flag.

Reviewer: ```suggestion
Apart from calling ADO API to trigger image build, Image Builder also supports preview mode. In preview mode,
Image Builder does not trigger the ADO pipeline but generates a YAML file with the pipeline definition. 
Using this mode allows for the validation of the pipeline definition syntax before triggering it. To use preview mode, add the `--ado-preview-run=true` flag.
```

> To specify a path to yaml file with pipeline definition use `--ado-preview-run-yaml-path` flag.

Reviewer: ```suggestion
To specify a path to the YAML file with the pipeline definition, use the `--ado-preview-run-yaml-path` flag.
```

> ## Deprecated features

Reviewer: ```suggestion
## Deprecated Features
```

> ### Build multi-architecture images

Reviewer: ```suggestion
### Build Multi-Architecture Images
```

With the introduction of the experimental BuildKit support, the tool now supports the repeatable flag `--platform`.
You can define multiple platforms you want to build an image for.

You can use all platforms supported by [BuildKit](https://github.com/moby/buildkit/blob/master/docs/multi-platform.md).

If you want to use experimental features, there is a new image with the tag suffix `-buildkit`.

### Build multiple variants of the same image

With `image-builder`, you can reuse the same `Dockerfile` to concurrently build different variants of the same image.
To predefine a set of the same `ARG` substitutions with different values, store them in the `variants.yaml` file .
Use that feature when you need to build an image with different versions of the same binary, for example, for different
versions of Kubernetes or Go.

The file has a simple structure:

```yaml
'main':
  KUBECTL_VERSION: "1.24.4"
'1.23':
  KUBECTL_VERSION: "1.23.9"
```

To use this feature, make sure that:

* you have the `variants.yaml` file in the **same directory** as the `Dockerfile`
* your `Dockerfile` contains `ARG` directives which are named after keys in `variants.yaml`

> ### Environment variables file

Reviewer: ```suggestion
### Environment Variables File
```

> `-env-file` specify path to the file with environment variables to be loaded in build. This flag is deprecated.

Reviewer: ```suggestion
`-env-file` specifies the path to the file with environment variables to be loaded in the build. This flag is deprecated.
```
